sdk/monitor/azure-monitor-query/samples/notebooks/sample_large_query.ipynb

```diff
@@ -0,0 +1,606 @@
+    "    try:\n",
+    "        response = await client.query_workspace(\n",
+    "            workspace_id=LOGS_WORKSPACE_ID,\n",
+    "            query=find_batch_endpoints_query,\n",
+    "            timespan=(start_time, end_time)\n",
+    "        )\n",
+    "    except HttpResponseError as e:\n",
+    "        print(\"Error batching endpoints by byte size: \", e)\n",
+    "\n",
+    "    batch_endpoints = [end_time]\n",
+    "    batch_endpoints += [row[0] for row in response.tables[0].rows]\n",
+    "    return batch_endpoints"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, define a function that will asynchronously execute a given query over a time range specified by a given start time and end time. This function will call the `query_workspace` method of the `LogsQueryClient` but will also handle retries in case of failures."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "async def execute_query(\n",
+    "    query: str, \n",
+    "    start_time: datetime, \n",
+    "    end_time: datetime, \n",
+    "    *, \n",
+    "    retries: int = 1,\n",
+    "    query_id: str = \"\",\n",
+    "    correlation_request_id: str = \"\",\n",
+    "):\n",
+    "    \"\"\"\n",
+    "    Asynchronously execute the given query, restricted to the given time range, and parse the API response.\n",
+    "\n",
+    "    :param str query: The query to execute.\n",
+    "    :param datetime start_time: The start of the time range to query.\n",
+    "    :param datetime end_time: The end of the time range to query.\n",
+    "    :keyword int retries: Number of times to retry the query if it fails.\n",
+    "    :keyword str query_id: Optional identifier for the query, used for printing.\n",
+    "    :keyword str correlation_request_id, Optional correlation ID to use in the query headers for tracing.\n",
+    "    \"\"\"\n",
+    "    headers = {}\n",
+    "    if correlation_request_id:\n",
+    "        headers[\"x-ms-correlation-request-id\"] = correlation_request_id\n",
+    "\n",
+    "    for i in range(retries + 1):\n",
+    "        if i > 0:\n",
+    "            print(f\"Attempt {i + 1} of {retries + 1}.\")\n",
+    "\n",
+    "        try:\n",
+    "            response = await client.query_workspace(\n",
+    "                workspace_id=LOGS_WORKSPACE_ID,\n",
+    "                query=query,\n",
+    "                timespan=(start_time, end_time),\n",
+    "                server_timeout=360,                 \n",
+    "                include_statistics=False, # Can be used for debugging.\n",
+    "                headers=headers\n",
+    "            )\n",
+    "        except HttpResponseError as e:\n",
+    "            print(f\"Error when attempting query {query_id} (query time span: {start_time} to {end_time}):\\n\\t\", e)\n",
+    "            if i < retries:\n",
```

Comment:
Why are we retrying outside of the library method? In general, we want to pass in retry information to the method itself (e.g. max_retries to the client.query_workspace method). Double retry loops tend to lead to very long and (for the user) unpredictable waits...